Heur sw shunts

TamerSobhy opened the pull request:
This PR addresses automatic switching shunts. It uses a similar approach to the one used by PSSE and OpenDSS with some modifications. 

Basically, all shunts are fixed (not variables). Solve PF (all iterations) then check voltage and adjust shunts. 

To avoid oscillations, I added a counter to lock the shunt if it moves in opposite directions. Below are the results of PSSE followed by the results of GRIDOPT (test case ACTIVSg10k in GAT/data).  In PSSE, 4 shunts are oscillating and the max number of iterations is reached. In GRIDOPT, the shunts are locked. 

If you feel we should write the code in C for speed purposes, let me know and I will do it. 

Best

____________________ PSSE Results _________________________

After iteration 2

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  13427     [MILTON FRE 1138.00]     0.00   -14.00
  30246     [BOULDER CI 313.800]    20.00    30.00
  50070     [TROPIC 1    161.00]    -8.00    -7.00
  50268     [LOA 1       161.00]   -38.00     0.00
  50279     [ESCALANTE 1 161.00]   -42.00     0.00

 5 switched shunts adjusted

After iteration 5

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  30246     [BOULDER CI~313.800]    30.00    20.00
  50070     [TROPIC 1    161.00]    -7.00    -8.00
  50268     [LOA 1       161.00]     0.00   -38.00
  50279     [ESCALANTE 1 161.00]     0.00   -42.00

 4 switched shunts adjusted

After iteration 8

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  30246     [BOULDER CI~313.800]    20.00    30.00
  50070     [TROPIC 1    161.00]    -8.00    -7.00
  50268     [LOA 1       161.00]   -38.00     0.00
  50279     [ESCALANTE 1 161.00]   -42.00     0.00

 4 switched shunts adjusted

After iteration 11

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  30246     [BOULDER CI~313.800]    30.00    20.00
  50070     [TROPIC 1    161.00]    -7.00    -8.00
  50268     [LOA 1       161.00]     0.00   -38.00
  50279     [ESCALANTE 1 161.00]     0.00   -42.00

 4 switched shunts adjusted

After iteration 14

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  30246     [BOULDER CI~313.800]    20.00    30.00
  50070     [TROPIC 1    161.00]    -8.00    -7.00
  50268     [LOA 1       161.00]   -38.00     0.00
  50279     [ESCALANTE 1 161.00]   -42.00     0.00

 4 switched shunts adjusted

After iteration 17

 SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
  30246     [BOULDER CI~313.800]    30.00    20.00
  50070     [TROPIC 1    161.00]    -7.00    -8.00
  50268     [LOA 1       161.00]     0.00   -38.00
  50279     [ESCALANTE 1 161.00]     0.00   -42.00

 4 switched shunts adjusted

After iteration 20

 Terminated after 20 iterations--reached iteration limit

 Largest mismatch:      0.01 MW      3.10 Mvar      3.10 MVA at bus 50395 [CASTLE DAL~2500.00]
 System total absolute mismatch:                   18.73 MVA

 SWING BUS SUMMARY:
   BUS#-SCT X-- NAME --X BASKV      PGEN     PMAX    PMIN      QGEN     QMAX    QMIN
  40845     PHOENIX 74 618.000    1398.2   1403.2   446.2     142.2    516.4  -115.1


_______________________ Results of GRIDOPT ______________________________________

Solver: NR

After iteration 9

SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
   50279     [   ESCALANTE 1    ]   0.00    -42.00 
   30246     [   BOULDER CI~3   ]  20.00    30.00  
   13427     [   MILTON FRE~1   ]   0.00    -14.00 

After iteration 13

SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
   50268     [      LOA 1       ]  -38.00    0.00  
   50070     [     TROPIC 1     ]  -8.00    -7.00  
   13427     [   MILTON FRE~1   ]   0.00    -14.00 

After iteration 16

SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
   13427     [   MILTON FRE~1   ]   0.00    -14.00 


SWITCHED SHUNTS LOCKED DUE TO OSCILLATIONS:
 X----------- AT BUS ----------X     Q 
   50279     [   ESCALANTE 1    ]   0.00  
   50268     [      LOA 1       ]  -38.00 
   50070     [     TROPIC 1     ]  -8.00  
   30246     [   BOULDER CI~3   ]  20.00  

After iteration 18

SWITCHED SHUNTS ADJUSTED:
 X----------- AT BUS ----------X    OLD Q    NEW Q
   13427     [   MILTON FRE~1   ]   0.00    -14.00 


SWITCHED SHUNTS LOCKED DUE TO OSCILLATIONS:
 X----------- AT BUS ----------X     Q 
   50279     [   ESCALANTE 1    ]   0.00  
   50268     [      LOA 1       ]  -38.00 
   50070     [     TROPIC 1     ]  -8.00  
   30246     [   BOULDER CI~3   ]  20.00

awig commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -706,6 +703,12 @@ def solve_problem(self, net, problem, save_problem=False, update_net=False):
 
+        shunt_actions = {sh.uid: [0, 0, sh.b] for sh in net.shunts if sh.is_in_service() and sh.is_switched_v()}
+        sh_reg_buses = [bus for bus in net.buses
+                        if bus.is_in_service() and
+                        len(bus.reg_shunts) > 0 and
+                        bus.v_max_reg > bus.v_min_reg and
+                        len([sh for sh in bus.reg_shunts if sh.is_switched_v() and sh.is_in_service()]) > 0]
```
Just a comment, but feels like we could have some shorthand functions to check for different regulating devices ``Shunt.is_valid_regulating()`` like gens or switched shunts have proper regulation or maybe ``Bus.get_valid_regulating_xxxx()``.  Seems we end up having to have all these checks a fair number of times.

awig commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -720,12 +723,159 @@ def c1(s):
+                    if len(reg_shunts) == 0:
+                        continue
+
+                    bus_v = s.get_primal_variables()[bus.index_v_mag]
```
Obviously this isn't the most efficient. Ideally we could have a function in pfnet to get this by index but seeing as how that isn't available. I think we should store the primal variables in a list outside of the loop and then index over that list. 

```
xhat = s.get_primal_variables()

for bus in sh_reg_buses:
   ...
   bus_v = xhat[bus.index_v_mag]
```

awig commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -720,12 +723,159 @@ def c1(s):
+                    shunt_mode == self.CONTROL_MODE_REG and
+                    solver.get_status() == 'solved'):
+
+                out = 0
```
or wait is this the number of active shunts (not locked)?

TamerSobhy replied:
This is the total number of sh adjustments. if it is zero, then the problem is solved. The check is on OPTALG https://github.com/romcon/OPTALG/pull/19/files

awig replied:
Oh i didn't realize there was associated changes in optalg

awig commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -720,12 +723,159 @@ def c1(s):
+                        if caps_on:
+                            # Turn off caps first
+                            for sh in reg_shunts:
+                                if s.swsh_call >= 3:
```
3 is the limit of the number of consecutive adjustments in order to avoid oscillations? 

Some ideas:
- prefer that this is a constant defined at the top of the file ``SHUNT_MODE_NR_LIMIT``
- should this be a parameter such as ``shunt_mode_nr_limit`` ?

awig commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -720,12 +723,159 @@ def c1(s):
+
+                        # Turn on Reactors
+                        for sh in reg_shunts:
+                            if s.swsh_call >= 3:
```
same as above?

TamerSobhy commented on gridopt/power_flow/ac_pf.py:
```diff
@@ -720,12 +723,159 @@ def c1(s):
+                            sh.b = sh.b_values[idx]
+                            out += 1
+                            shunt_actions[sh.uid][0] += 1
+                            break
```
Ideally, shunt values are sorted when reading raw files. this break here moves shunts one step at a time. If all steps are switched on, then smaller steps will be switched off next.
This is similar to PSSE/OpenDSS. We can try other techniques as well. something similar to using dv/dq to determine the optimal combinations to be switched, but it will slow down the computations a bit.

TamerSobhy commented:
Thanks @awig, Do you want me to update the files and add the test case?

TamerSobhy commented:
@awig 
I updated the code according to your suggestions. I also updated the code to print the shunt switched after each iteration without repeating printing the same message. 
All other callbacks should be evaluated if the solver status != 'solved'. I added this line to avoid double calls.
I added a test case comparing the results.
